### PR TITLE
feat invariant (#5868) - configure calldata fuzzed addresses dictionary

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation

Atm there's no way to configure calldata fuzzed params behavior - this could in particular be an issue when fuzzing addresses as there's little to no chance the same address is reused.
For invariant testing this is a bigger issue as one would often want to randomly reuse addresses when calling different handler selectors (see https://github.com/foundry-rs/foundry/issues/5868). Some projects maintains an array of addresses in handlers and fuzz the indexes / prank with users from array of addresses - that's a little bit hard to mantain and cannot scale for too many addresses.

## Solution

This PR introduce an additional config for the dictionary of addresses used in calldata fuzzing.
```
[invariant]
max_calldata_fuzz_dictionary_addresses=100
```

- added `FuzzDictionaryConfig.max_calldata_fuzz_dictionary_addresses` option to specify how many random addresses to generate and to randomly select from when fuzzing calldata. If option is not specified then current behavior applies
- to narrow down number of runs / addresses involved in invariant test the `CalldataFuzzDictionaryConfig` is populated with random addresses plus all accounts from db (from `EvmFuzzState`)
- added `fuzz_calldata_with_config` fn that accepts `Option<CalldataFuzzDictionary>` as param. Non invariants tests use existing `fuzz_calldata` fn and pass None as config arg

Further improvements:
- add other options to generate fuzz dictionary address (like from external csv file, from inline config, etc)
- extend config also for basic (non invariant) tests
- extend calldata fuzz config with rules for other fuzzed param types  

Todo:
- add inline config
- add tests

### CODE REVIEW

File: crates/evm/fuzz/src/strategies/calldata.rs [hunk 1]
```diff
@@ -1,18 +1,65 @@
...
+}
+
+impl fmt::Debug for CalldataFuzzDictionaryConfig {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("CalldataFuzzDictionaryConfig").field("addresses", &self.addresses).finish()
+    }
+}
+
+impl CalldataFuzzDictionaryConfig {
+    pub fn new(config: &FuzzDictionaryConfig, state: EvmFuzzState) -> Self {
```
Comment (mattsse) [inline crates/evm/fuzz/src/strategies/calldata.rs:25]: please add some docs

File: crates/evm/fuzz/src/strategies/calldata.rs [hunk 2]
```diff
@@ -1,18 +1,65 @@
...
 use alloy_dyn_abi::JsonAbiExt;
 use alloy_json_abi::Function;
-use alloy_primitives::Bytes;
+use alloy_primitives::{Address, Bytes};
+use foundry_config::FuzzDictionaryConfig;
+use hashbrown::HashSet;
 use proptest::prelude::{BoxedStrategy, Strategy};
+use std::{fmt, sync::Arc};
+
+pub type CalldataFuzzDictionary = Arc<CalldataFuzzDictionaryConfig>;
```
Comment (mattsse) [inline crates/evm/fuzz/src/strategies/calldata.rs:10]: I really disklike this alias style, this makes it much harder to navigate.

I know we have this in a few other places, but this isn't very helpful. I'd prefer a new wrapper type like

```rust
struct CalldataFuzzDictionary {
   inner: Arc<CalldataFuzzDictionaryConfig>
}
```

**Review by mattsse — CHANGES_REQUESTED** [review]
I think that makes sense

left some nits

**Comment by aviggiano** [comment]
Will this also allow users to specify the number of distinct senders? 

AFAIK Foundry currently uses an unlimited number of `msg.sender`, which is not always ideal in some scenarios of invariant testing. Echidna/Medusa, for example, uses only 3 addresses by default. 

It should be nice to configure that as well

**Comment by grandizzy** [comment]
> Will this also allow users to specify the number of distinct senders?
> 
> AFAIK Foundry currently uses an unlimited number of `msg.sender`, which is not always ideal in some scenarios of invariant testing. Echidna/Medusa, for example, uses only 3 addresses by default.
> 
> It should be nice to configure that as well

correct, there is a new config `max_calldata_fuzz_dictionary_addresses` to specify the number of additional random addresses when fuzzing calldata.
Re `msg.sender` - that's already finite to the number of accounts from state and you can get to your desired set  by using `targetSender(address newTargetedSender_)` see options here https://book.getfoundry.sh/forge/invariant-testing#invariant-test-helper-functions

File: crates/evm/fuzz/src/strategies/calldata.rs (hunk 2 above)
Comment (grandizzy) [reply crates/evm/fuzz/src/strategies/calldata.rs]: done with https://github.com/foundry-rs/foundry/pull/7240/commits/b77dcb6fca13e732067cf50ac517e6c5fc641122 plus some code cleanup

File: crates/evm/fuzz/src/strategies/calldata.rs (hunk 1 above)
Comment (grandizzy) [reply crates/evm/fuzz/src/strategies/calldata.rs:44]: added in https://github.com/foundry-rs/foundry/pull/7240/commits/b77dcb6fca13e732067cf50ac517e6c5fc641122

**Review by mattsse — APPROVED** [review]
lgtm, pending conflicts

**Comment by grandizzy** [comment]
> lgtm, pending conflicts

done

**Pull request merged**